ITL-6 Add sorting functionality

EmanuelRusuu opened the pull request.

EmanuelRusuu committed:
ITL-24 Make a component and six buttons first

- edited the filtering logic

EmanuelRusuu committed:
ITL-25 Implement the ability to sort by title

- sorting by title 
- sorting by description (both in ascending order, as default)

EmanuelRusuu committed:
ITl-26 Implement the ability to sort by due date

- sorting by date

EmanuelRusuu committed:
ITL-27 Implement the ability to sort by priority

- sorting by priority

EmanuelRusuu committed:
ITL-28 Implement the ability to sort in ascending order

EmanuelRusuu committed:
ITL-29 Implement the ability to sort in descending order

- user can sort the todos ascending/descending by title, description, priority or date

EmanuelRusuu committed:
[ITL-6] Fixed issues

- improved the code to being more efficient,
- fixed the deleting of todos by adding a unique id for every todo,
- implemented searching and sorting for done Todos array
- search now works only after clicking the search button,
- removed maxlimit on the inputs,
- debounce for searching todos

kincsokiss commented on todoapp/src/components/FinishedTodos/FinishedTodos.vue:
```diff
@@ -0,0 +1,116 @@
+      )
+  } else if (selectedButton === 'date') {
+    return props.searchFinishedTodos.slice().sort((a, b) => {
+      const dateA = new Date(
```
Since the new Date(...) are almost the same, please create a local function which return a Date based on the sent `a` or `b` parameter in order not to duplicate the code.

kincsokiss commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,243 @@
+  isEditing: false
+}
+
+function addTodo() {
```
The `addTodo` and `addTodo` functions are almost the same. Please create general function for the common code s.

kincsokiss commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,243 @@
+}
+
+function createId() {
+  const timestamp = Date.now()
```
Please decide between moment JS and JS date, and use only one from them.

kincsokiss commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,243 @@
+  if (todo.isChecked) {
+    const updatedFinishedTodos = [...finishedTodos.value, todo]
+    const updatedTodos = todos.value.filter((item) => item.id !== todo.id)
+    finishedTodos.value = updatedFinishedTodos
```
Please move this part after the if ..else, this block added to if and else as well, so you don't need to duplicate it.
`finishedTodos.value = updatedFinishedTodos; todos.value = updatedTodos`

kincsokiss commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,243 @@
+}
+
+function updateTodo(updatedTodo: TodoType) {
+  const todoIndex = todos.value.findIndex((todo) => todo.id === updatedTodo.id)
```
Please create a general local function to return an index from an array and use it in the whole file, because the same findIndex is implemented more times.

kincsokiss commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,243 @@
+  currentEditedTodo.value = null
+}
+
+const searchTodos = computed(() => {
```
Please create a general function for `searchTodos` and `searchFinishedTodos ` and add the needed arrays as the parameter.

kincsokiss commented on todoapp/src/components/Todos/Todo.vue:
```diff
@@ -0,0 +1,108 @@
+}>()
+
+function editTodo(id: number, isFinishedTodo: boolean) {
+  if (props.currentEditedTodo === null) {
```
You can use `!props.currentEditedTodo` instead of `props.currentEditedTodo === null` in the condition.

kincsokiss commented on todoapp/src/components/Todos/Todos.vue:
```diff
@@ -0,0 +1,124 @@
+      )
+  } else if (selectedButton === 'date') {
+    return props.searchTodos.slice().sort((a, b) => {
+      const dateA = new Date(
```
Please review this if..else statement and try to create general local functions for the code duplications.

kincsokiss replied:
The `Todos.vue` and `FinishedTodos.vue` are almost the same. You should create a general component for the common blocks and use it in the `Todos.vue` and `FinishedTodos.vue` instead of duplicating the logic.

EmanuelRusuu committed:
[ITL-6] Fixed PR issues

- created general / helper functions,
- used moment instead of JS Date,
- created general Todolist component.

gerebdavid commented on todoapp/src/components/FilterTodos.vue:
```diff
@@ -0,0 +1,57 @@
+        v-for="button in buttons"
+        :key="button.key"
+        class="sorting-btn border-black"
+        :class="{
```
Create a computed from this

gerebdavid commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,212 @@
+    class="relative flex flex-col items-center mt-32 max-w-288 w-4/5 xsm:max-w-500 xsm:w-11/12 xsm:mt-44 md:max-w-610 md:w-4/5 md:mt-48"
+  >
+    <Header @add-todo="addTodo" />
+    <SearchTodos
```
wrap the two components inside a <template> and add the v-if to the template

gerebdavid commented on todoapp/src/components/TodoApp.vue:
```diff
@@ -0,0 +1,212 @@
+}
+
+const searchTodos = computed(() => {
+  return filterTodosBySearchInput(todos.value, searchInput.value)
```
Use shorter version (delete return and {})
